Issue #3039084: Always show Full Name

OH17L opened the pull request:
## Problem
Can not see Full Name when turned on the Nickname field.

## Solution
1. Add permission to the social_profile_fields module to always show Full Name.
2. Add a setting to the social_profile_fields module to limit search and mention using "Full Names" to the permission to see real name.

## Issue tracker
- https://www.drupal.org/project/social/issues/3039084
- https://jira.goalgorilla.com/browse/GPDE-65

## How to test
- [ ] Turn on **Social Profile Fields** module.
- [ ] Log in as **SM** and navigate to **Manage > Configuration > Open Social Settings**.
- [ ] You should see unchecked **Limit search and mention** checkbox.
- [ ] Enable **Nickname** field.
- [ ] Create **First Test User** (username - **ftu**, first name - **First**, last name - **Test User**, nickname - **First User**).
- [ ] Create **Second Test User** (username - **stu**, first name - **Second**, last name - **Test User**, nickname - **Second User**).
- [ ] Create **Test Role** role and add **Always show Full Name** permission to this role.
- [ ] Assign **Test Role** role to **First Test User**.
- [ ] Log in as **Second Test User**.
- [ ] When you type **@ftu** in post field then you should see **First User** in drop-down list.
- [ ] Log in as **First Test User**.
- [ ] When you type **@stu** in post field then you should see **Second Test User** instead of **Second User** in drop-down list.
- [ ] Log in as **SM** and navigate to **Manage > Configuration > Open Social Settings**.
- [ ] Check **Limit search and mention** checkbox and press **Save configuration**.
- [ ] Log in as **Second Test User**.
- [ ] When you type **@First Test User** in post field then you should see empty drop-down list.
- [ ] When you use **First Test** as search phrase then you should not see **First Test User** user in search results.
- [ ] Log in as **First Test User**.
- [ ] When you type **@Second Test User** in post field then you should see **Second Test User** in a drop-down list.
- [ ] When you use **Second Test** as search phrase then you should see **Second Test User** user in search results.

## Release notes
A user can not use the full name of other users for mention them or search them when the **Limit search and mention** option (the **Social Profile Fields** module settings page) is activated and the user does not have the **Always show Full Name** permission.

ohorbatiuk committed:
Issue #3039084 by chmez: Create permission for always showing Full Name.

ohorbatiuk committed:
Issue #3039084 by chmez: Add new option on the settings page.

ronaldtebrake commented:
Should we set the default to false on install?

OH17L commented:
> Should we set the default to false on install?

No, we shouldn't. It will be false automatica

OH17L commented:
> Should we set the default to false on install?

Maybe...

ohorbatiuk committed:
Merge branch '8.x-4.x' of https://github.com/goalgorilla/open_social into feature/3039084

ohorbatiuk committed:
Issue #3039084 by chmez: Check if can using a full name for search for mentions.

ohorbatiuk committed:
Merge branch '8.x-4.x' of https://github.com/goalgorilla/open_social into feature/3039084

ohorbatiuk committed:
Issue #3039084 by chmez: Use shorter format for conditions for searching in a part of a field value.

ohorbatiuk committed:
Issue #3039084 by chmez: Turn off limitation by default.

ohorbatiuk committed:
Issue #3039084 by chmez: Fix error when searching users when profile tags are absent.

ohorbatiuk committed:
Issue #3039084 by chmez: Prohibit using the full name for search by users without permission when the limitation is turned on.

bramtenhove commented on modules/social_features/social_profile/src/SocialProfileTrait.php:
```diff
@@ -122,20 +159,26 @@ private function startQuery() {
    *   The select query.
    */
   private function sortQuery(SelectInterface $query, $name, $suggestion_format) {
```
Please add that to the `@param string $name` description that it is already sanitized.

OH17L replied:
Is it enought?

bramtenhove replied:
Yes, thanks!

ohorbatiuk committed:
Merge branch '8.x-4.x' of https://github.com/goalgorilla/open_social into feature/3039084

ohorbatiuk committed:
Issue #3039084 by chmez: Extend description for name argument of sorting method and delete sorting method name which using by default.

ohorbatiuk committed:
Issue #3039084 by chmez, kingdutch: Use longer name of configuration.

ohorbatiuk committed:
Issue #3039084 by chmez: Fix getting username instead of nickname when the full name is empty.

ohorbatiuk committed:
Issue #3039084 by chmez: Use single-line format for method for checking if can using full name.

bramtenhove commented on modules/social_features/social_profile/modules/social_profile_fields/social_profile_fields.module:
```diff
@@ -148,3 +155,54 @@ function social_profile_fields_social_user_export_plugin_info_alter(array &$plug
+/**
+ * Implements hook_search_api_query_alter().
+ */
+function social_profile_fields_search_api_query_alter(QueryInterface &$query) {
```
Good explanation. The same goes for usernames, everyone should always be able to find someone by their username. This is also how it works now afaik.

Kingdutch committed:
Add nickname field to search_all index

This is needed so the nickname can be used in search filter queries that
will be implemented for the privacy module. This should also improve
search_all search results.

Kingdutch committed:
Move first name and last name into search indices

This change moves the first name and last name of user profiles from
being indexed through the rendered entity into their own fields in the
index. This allows us to apply access control to the fields which is a
new feature that will be implemented in the social_profile_privacy
module.

Kingdutch committed:
Fix preprocessing for name fields in search

The fields need to be preprocessed so that they are properly tokenised
and work with non-ascii characters.

Kingdutch committed:
Filter out matches on full name for non-eligible users

When the users do not have the permission to see the full name and the
full name is protected then the matches in search that arrived from the
name and no other field are filtered out.

Kingdutch committed:
Move full name protection code to social_profile_privacy

This moves the code for hiding the full name of users to the
social_profile_privacy. It also modifies the `useFullName` condition to
work regardless of there being a nickname. This ensures that even if the
social_profile_fields module is not enabled then the Full Name can still
be hidden and the username used as display name.

The only issue remaining in this commit is that the full name hiding
happens in the social profile fields module. This will need a better
solution by moving the authoritative rule on this to the social user
module.

Kingdutch committed:
Add user display name resolution to social_user

Multiple modules were trying to overwrite the user display name. The
social_profile module wanted to display the user's full name. The
social_profile_fields module wanted a nickname to take precedence and
finally the social_profile_privacy wanted a say in when the full name
would be displayed.

To resolve this we centralise the username resolution in the social_user
module and provide other modules the ability to add suggestions using a
weight to specify order of preference. An alter hook is implemented to
give modules the chance to overwrite or unset suggestions. This can be
used for example to implement access checking.

If no suggestions are left then we simply do not alter the name and
default to Drupal's built-in behaviour.

Kingdutch committed:
Update full name hiding texts and explanations

This incorporates the feedback from Natasha.

Kingdutch committed:
Ensure allowed users see the full name when limiting is enabled

There was a slight error when the code was moved to the
social_profile_privacy module that caused users with the "View full
name" permission to still see the nickname instead of the full name when
user name limiting was enabled.

Kingdutch commented:
I'm closing this pull request. Given the changes that have been made this will have to be introduced in 8.x-5.x instead of 8.x-4.x
